x/qgb/abci.go

```diff
@@ -9,20 +9,37 @@ import (
 
 // EndBlocker is called at the end of every block
 func EndBlocker(ctx sdk.Context, k keeper.Keeper) {
-	handleValsetRequest(ctx, k)
 	handleDataCommitmentRequest(ctx, k)
+	handleValsetRequest(ctx, k)
 }
 
 func handleDataCommitmentRequest(ctx sdk.Context, k keeper.Keeper) {
-	if ctx.BlockHeight()%int64(types.DataCommitmentWindow) != 0 {
-		return
+	if ctx.BlockHeight() != 0 && ctx.BlockHeight()%int64(types.DataCommitmentWindow) == 0 {
+		dataCommitment, err := k.GetCurrentDataCommitment(ctx)
+		if err != nil {
+			panic(sdkerrors.Wrap(err, "coudln't get current data commitment"))
+		}
+		k.StoreAttestation(ctx, &dataCommitment)
+		k.SetLatestAttestationNonce(ctx, dataCommitment.Nonce)
 	}
-	k.SetDataCommitmentRequest(ctx)
 }
 
 func handleValsetRequest(ctx sdk.Context, k keeper.Keeper) {
-	//get the last valsets to compare against
-	latestValset := k.GetLatestValset(ctx)
+	// FIXME can this be done here?
+	// We need to initialize this value in the begining...
+	if !k.CheckLatestAttestationNonce(ctx) {
+		k.SetLatestAttestationNonce(ctx, 0)
+	}
+	// get the last valsets to compare against
+	var latestValset *types.Valset
+	if k.CheckLatestAttestationNonce(ctx) && k.GetLatestAttestationNonce(ctx) != 0 {
+		var err error
+		latestValset, err = k.GetLatestValset(ctx)
+		if err != nil {
+			panic(err)
```

Comment:
as per normal, we just have to document when this panic can occur. we can handle this either in this PR or in a different issue

same with line 75

Reply:
https://github.com/celestiaorg/celestia-app/issues/501